curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -0,0 +1,860 @@
+    info->set_poolname(pPool.GetName());
+    response->set_allocated_zoneinfo(info);
+}
+
+void TopologyManager::ListPoolZone(const ListPoolZoneRequest* request,
+                                   ListPoolZoneResponse* response) {
+    Pool pPool;
+    if (!topology_->GetPool(request->poolid(), &pPool)) {
+        response->set_statuscode(TopoStatusCode::TOPO_POOL_NOT_FOUND);
+        return;
+    }
+    std::list<ZoneIdType> zidList = pPool.GetZoneList();
+    response->set_statuscode(TopoStatusCode::TOPO_OK);
+    for (ZoneIdType id : zidList) {
+        Zone zone;
+        if (topology_->GetZone(id, &zone)) {
+            ZoneInfo *info = response->add_zones();
+            info->set_zoneid(zone.GetId());
+            info->set_zonename(zone.GetName());
+            info->set_poolid(pPool.GetId());
+            info->set_poolname(pPool.GetName());
+        } else {
+            LOG(ERROR) << "Topology has counter an internal error: "
+                       << "[func:] ListPoolZone, "
+                       << "[msg:] Zone not found, id = "
+                       << id;
+            response->set_statuscode(TopoStatusCode::TOPO_INTERNAL_ERROR);
+            return;
+        }
+    }
+}
+
+void TopologyManager::CreatePool(const CreatePoolRequest *request,
+                                        CreatePoolResponse *response) {
+    PoolIdType pid = topology_->AllocatePoolId();
+    if (pid == static_cast<PoolIdType>(UNINTIALIZE_ID)) {
+        response->set_statuscode(TopoStatusCode::TOPO_ALLOCATE_ID_FAIL);
+        return;
+    }
+
+    Pool::RedundanceAndPlaceMentPolicy rap;
+    if (!Pool::TransRedundanceAndPlaceMentPolicyFromJsonStr(
+        request->redundanceandplacementpolicy(), &rap)) {
+        LOG(ERROR) << "[TopologyManager::CreatePool]:"
+                << "parse redundanceandplacementpolicy fail.";
+        response->set_statuscode(TopoStatusCode::TOPO_INVALID_PARAM);
+        return;
+    }
+
+    uint64_t time = TimeUtility::GetTimeofDaySec();
+
+    Pool pool(pid, request->poolname(), rap, time, true);
+
+    TopoStatusCode errcode = topology_->AddPool(pool);
+    if (TopoStatusCode::TOPO_OK == errcode) {
+        response->set_statuscode(errcode);
+        PoolInfo *info = new PoolInfo();
+        info->set_poolid(pid);
+        info->set_poolname(request->poolname());
+        info->set_createtime(time);
+        info->set_redundanceandplacementpolicy(
+            pool.GetRedundanceAndPlaceMentPolicyJsonStr());
+        response->set_allocated_poolinfo(info);
+    } else {
+        response->set_statuscode(errcode);
+    }
+}
+
+void TopologyManager::DeletePool(const DeletePoolRequest *request,
+                                        DeletePoolResponse *response) {
+    Pool pool;
+    if (!topology_->GetPool(request->poolid(), &pool)) {
+        response->set_statuscode(TopoStatusCode::TOPO_POOL_NOT_FOUND);
+        return;
+    }
+
+    TopoStatusCode errcode = topology_->RemovePool(pool.GetId());
```

Comment:
remove pool should judge copyset in the pool is clear or not.

Reply:
> remove pool should also clear copyset and partition

The condition of remove pool is the pool is empty(have no zone, no servers in zone, no metaservers in server). If the condition is ok, will remove the pool. But have no remove copyset interface now, I will write a TODO.